Zend/zend_compile.c

```diff
@@ -6894,19 +6916,21 @@ void zend_compile_class_decl(znode *result, zend_ast *ast, zend_bool toplevel) /
 			zend_error_noreturn(E_COMPILE_ERROR, "Constructor %s::%s() cannot be static",
 				ZSTR_VAL(ce->name), ZSTR_VAL(ce->constructor->common.function_name));
 		}
-		if (ce->constructor->common.fn_flags & ZEND_ACC_HAS_RETURN_TYPE) {
+		if (ce->constructor->common.fn_flags & ZEND_ACC_HAS_RETURN_TYPE
+			&& ZEND_TYPE_PURE_MASK(ce->constructor->common.arg_info[-1].type) != MAY_BE_VOID) {
 			zend_error_noreturn(E_COMPILE_ERROR,
-				"Constructor %s::%s() cannot declare a return type",
+				"Constructor %s::%s() must return void",
```

Comment:
I agree with Danack that this message is a bit confusing. For me, it indicates that the problem is related to the return value, while in fact it's a compile-time error, related to the return type. Therefore, I'd prefer a message similar to the following ones:
`%s::%s(): Return type can only be void` or `%s::%s(): Return type must be void when declared`.

Please note that the previous format is in line what we currently use for return value errors, e.g.:
`foo(): Return value must be of type array, null returned`

But probably error messages are the least important part of an RFC when in its infancy :)

Reply:
For consistency, I will wait until Gabriel finalizes the wording he uses on his PR (since it's `must return void` at the moment).

Reply:
Yes, that makes sense! First, we should come up with a consistent format for all the magic method signature issues.

Reply:
Resolved.